src/map.cpp

```diff
@@ -755,46 +592,253 @@ const vehicle *map::vehproceed()
+    for( const auto &ps : boarded ) {
+        player *psg = veh.get_passenger( ps );
+        if( psg == nullptr ) {
+            debugmsg( "throw passenger: empty passenger at part %d", ps );
+            continue;
+        }
+
+        const tripoint part_pos = pt + veh.parts[ps].precalc[0];
+        if( psg->pos() != part_pos ) {
+            debugmsg( "throw passenger: passenger at %d,%d,%d, part at %d,%d,%d",
+                psg->posx(), psg->posy(), psg->posz(), part_pos.x, part_pos.y, part_pos.z );
+            veh.parts[ps].remove_flag( vehicle_part::passenger_flag );
+            continue;
+        }
+
+        bool throw_from_seat = false;
+        if( veh.part_with_feature( ps, VPFLAG_SEATBELT ) == -1 ) {
+            throw_from_seat = d_vel * rng( 80, 120 ) / 100 > ( psg->str_cur * 1.5 + 5 );
+        }
+
+        // Damage passengers if d_vel is too high
+        if( d_vel > 60 * rng( 50,100 ) / 100 && !throw_from_seat ) {
+            const int dmg = d_vel / 4 * rng( 70,100 ) / 100;
+            psg->hurtall( dmg, nullptr );
+            psg->add_msg_player_or_npc( m_bad,
+                _("You take %d damage by the power of the impact!"),
+                _("<npcname> takes %d damage by the power of the impact!"),  dmg );
+        }
+
+        if( veh.player_in_control( *psg ) ) {
+            const int lose_ctrl_roll = rng( 0, d_vel );
+            if( lose_ctrl_roll > psg->dex_cur * 2 + psg->skillLevel("driving") * 3 ) {
+                add_msg( m_warning,
+                    _("You lose control of the %s."),
+                    _("<npcname> loses control of the %s."),
+                    veh.name.c_str() );
+                int turn_amount = (rng(1, 3) * sqrt((double)abs( veh.velocity ) ) / 2) / 15;
+                if( turn_amount < 1 ) {
+                    turn_amount = 1;
+                }
+                turn_amount *= 15;
+                if( turn_amount > 120 ) {
+                    turn_amount = 120;
+                }
+                coll_turn = one_in( 2 ) ? turn_amount : -turn_amount;
+            }
+        }
+
+        if( throw_from_seat ) {
+            psg->add_msg_player_or_npc(m_bad,
+                _("You are hurled from the %s's seat by the power of the impact!"),
+                _("<npcname> is hurled from the %s's seat by the power of the impact!"),
+                veh.name.c_str());
+            unboard_vehicle( part_pos );
+            g->fling_creature(psg, direction + rng(0, 60) - 30,
+                ( d_vel - psg->str_cur < 10 ) ? 10 : d_vel - psg->str_cur );
+        }
+    }
+
+    return coll_turn;
+}
+
+float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
+                                      const std::vector<veh_collision> &collisions )
+{
+    // Effects of colliding with another vehicle:
+    //  transfers of momentum, skidding,
+    //  parts are damaged/broken on both sides,
+    //  remaining times are normalized
+    const veh_collision &c = collisions[0];
```

Comment:
What if `collisions` has several entries? The calling code goes through all of them and calls this function several times, but this will always use the firsts entry, which may not match.

This could probably lead to problems if the first entry is a vehicle with many parts and the part index is huge, but the next colliding vehicle (`veh2`) in the next call to this function has only few parts and the index from `c.target_part` is not valid there.